lib/luaxtable.c

```diff
@@ -0,0 +1,232 @@
+* "Software"), to deal in the Software without restriction, including
+* without limitation the rights to use, copy, modify, merge, publish,
+* distribute, sublicense, and/or sell copies of the Software, and to
+* permit persons to whom the Software is furnished to do so, subject to
+* the following conditions:
+*
+* The above copyright notice and this permission notice shall be
+* included in all copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
+#include <linux/module.h>
+#include <linux/version.h>
+#include <linux/types.h>
+#include <linux/netfilter.h>
+#include <linux/netfilter/x_tables.h>
+
+#include <lua.h>
+#include <lauxlib.h>
+#include <lunatik.h>
+
+typedef enum luaxtable_type_e {
+	LUAXTABLE_TMATCH,
+	LUAXTABLE_TTARGET,
+} luaxtable_type_t;
+
+typedef struct luaxtable_s {
+	lunatik_object_t *runtime;
+	union {
+		struct xt_match match;
+		struct xt_target target;
+	};
+	luaxtable_type_t type;
+	bool registered;
+} luaxtable_t;
+
+static bool luaxtable_match(const struct sk_buff *skb, struct xt_action_param *par)
+{
+	return true;
+}
+
+static int luaxtable_match_check(const struct xt_mtchk_param *par)
+{
+	return 0;
+}
+
+static void luaxtable_match_destroy(const struct xt_mtdtor_param *par)
+{
+}
+
+static unsigned int luaxtable_target(struct sk_buff *skb, const struct xt_action_param *par)
+{
+	return NF_ACCEPT;
+}
+
+static int luaxtable_target_check(const struct xt_tgchk_param *par)
+{
+	return 0;
+}
+
+static void luaxtable_target_destroy(const struct xt_tgdtor_param *par)
+{
+}
+
+static void luaxtable_release(void *private);
+
+static const luaL_Reg luaxtable_mt[] = {
+	{"__gc", lunatik_deleteobject},
+	{NULL, NULL}
+};
+
+static const lunatik_class_t luaxtable_class = {
+	.name = "xtable",
+	.methods = luaxtable_mt,
+	.release = luaxtable_release,
+	.sleep = false,
+};
+
+#define luaxtable_setinteger(L, idx, hook, field) 		\
+do {								\
+	lunatik_checkfield(L, idx, #field, LUA_TNUMBER);	\
+	hook->field = lua_tointeger(L, -1);			\
+	lua_pop(L, 1);						\
+} while (0)
+
+#define luaxtable_setstring(L, idx, hook, field, maxlen)        \
+do {								\
+	const char *str;					\
+	size_t len;						\
+								\
+	lunatik_checkfield(L, idx, #field, LUA_TSTRING);	\
+	str = lua_tolstring(L, -1, &len);			\
+	if (len >= maxlen)					\
+		luaL_error(L, "'%s' is too long", #field);	\
+	strncpy((char *)hook->field, str, maxlen);		\
+	lua_pop(L, 1);						\
```

Comment:
```suggestion
	size_t len;						\
	lunatik_checkfield(L, idx, #field, LUA_TSTRING);	\
	const char *str = lua_tolstring(L, -1, &len);		\
	if (len >= maxlen)					\
		luaL_error(L, "'%s' is too long", #field);	\
	strncpy((char *)hook->field, str, maxlen);		\
	lua_pop(L, 1);						\
```

as you are using C99 style declarations (scattered around the block, not only at the beginning), you can do it here as well.. I'm not against it.. but I think we should apply to the other modules to standardize the code base.. perhaps we should create a ticket for this..

we should use `maxlen - 1` (for `\0`) on strncpy, right? Perhaps a better idea is to pass `XT_EXTENSION_MAXNAMELEN - 1` to this function and do `if (len > maxlen)`... it might be clearer.. so we use `len` as in the [libc semantics](https://pubs.opengroup.org/onlinepubs/9699919799/functions/strlen.html)..